src/Symfony/Component/Security/Http/EventListener/VerifyAuthenticatorCredentialsListener.php

```diff
@@ -0,0 +1,78 @@
+<?php
+
+namespace Symfony\Component\Security\Http\EventListener;
+
+use Symfony\Component\EventDispatcher\EventSubscriberInterface;
+use Symfony\Component\Security\Core\Encoder\EncoderFactoryInterface;
+use Symfony\Component\Security\Core\Exception\BadCredentialsException;
+use Symfony\Component\Security\Core\Exception\LogicException;
+use Symfony\Component\Security\Http\Authenticator\CustomAuthenticatedInterface;
+use Symfony\Component\Security\Http\Authenticator\PasswordAuthenticatedInterface;
+use Symfony\Component\Security\Http\Authenticator\TokenAuthenticatedInterface;
+use Symfony\Component\Security\Http\Event\VerifyAuthenticatorCredentialsEvent;
+
+/**
+ * This listeners uses the interfaces of authenticators to
+ * determine how to check credentials.
+ *
+ * @author Wouter de Jong <wouter@driveamber.com>
+ *
+ * @final
+ * @experimental in 5.1
+ */
+class VerifyAuthenticatorCredentialsListener implements EventSubscriberInterface
+{
+    private $encoderFactory;
+
+    public function __construct(EncoderFactoryInterface $encoderFactory)
+    {
+        $this->encoderFactory = $encoderFactory;
+    }
+
+    public function onAuthenticating(VerifyAuthenticatorCredentialsEvent $event): void
+    {
+        if ($event->areCredentialsValid()) {
+            return;
+        }
+
+        $authenticator = $event->getAuthenticator();
+        if ($authenticator instanceof PasswordAuthenticatedInterface) {
+            // Use the password encoder to validate the credentials
+            $user = $event->getUser();
+            $presentedPassword = $authenticator->getPassword($event->getCredentials());
+            if ('' === $presentedPassword) {
```

Comment:
Hmm, this is another spot where the "opt out" is built into the method itself (i.e. if `getPassword()` returns null, don't do validation). Again, I think that's too dangerous :/. If we need an opt-out, it needs to be a different method. Maybe, like with CSRF, we have a `shouldValidatePassword(): bool` method (again, maybe we pass the `Request` as an argument?)

Reply:
marking as resolved, the badge system (https://github.com/symfony/symfony/pull/33558#issuecomment-610111667) will remove this interface